runtime/lua/vim/lsp/util.lua

```diff
@@ -1240,45 +1230,40 @@ local function find_window_by_var(name, value)
   end
 end
 
---- Trims empty lines from input and pad top and bottom with empty lines
----
----@param contents table of lines to trim and pad
----@param opts table with optional fields
----             - pad_top    number of lines to pad contents at top (default 0)
----             - pad_bottom number of lines to pad contents at bottom (default 0)
----@return table table of trimmed and padded lines
-function M._trim(contents, opts)
-  validate({
-    contents = { contents, 't' },
-    opts = { opts, 't', true },
-  })
-  opts = opts or {}
-  contents = M.trim_empty_lines(contents)
-  if opts.pad_top then
-    for _ = 1, opts.pad_top do
-      table.insert(contents, 1, '')
-    end
+---Replaces separator lines by the given divider and removing surrounding blank lines.
+---@param contents string[]
+---@param divider string
+---@return string[]
+local function replace_separators(contents, divider)
+  ---@param line string
+  local function is_blank_line(line)
```

Comment:
I would move `is_blank_line()`, `is_separator_line()` and `get_markdown_fences()` out from being closures.

If you want to reduce scope of certain functions/state, then better put all the markdown stuff into a separate module.

Reply:
Follow up PR.

I don't mind closures or separate functions, as long as they are local and not exported. 

Having them as closures makes it much easier to move to a different module later, though.

Reply:
What I've seen happen (many times), is people will write small little helper closures with no state. Move on. Then other contributors start chiming in with _features_ that exploit the fact these are closures and begin adding upvals. Now you can't easily refactor. This is sort of what's already happened in this file.

I've already spent some time unwinding a lot of this kind of stuff in nvim-lspconfig which has full on closure classes.

By making a proactive step to fix a structure a little bit now will mitigate against these risks by enforcing a stronger functional style.

Reply:
Sorry, but that is abstract worry blocking concrete improvements. And again, this function is slated to be moved, so portability is a boon.

Reply:
I moved the closures anyway. I don't want to ruin @lewis6991's weekend.